compat/optional.h

```diff
@@ -6,8 +6,16 @@
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
 
-#ifndef SQUID__SRC_BASE_OPTIONAL_H
-#define SQUID__SRC_BASE_OPTIONAL_H
+#ifndef SQUID__COMPAT_OPTIONAL_H
+#define SQUID__COMPAT_OPTIONAL_H
+
+#if __cplusplus
+#if __has_include(<optional>)
+
+#include <optional>
+template<class T> using Optional = std::optional<T>;
```

Comment:
:question: I am concerned that doing this will introduce difficult to triage runtime bugs where our buggy[^1] Optional code in the reporter environment is substituted by correct std::optional implementation in the developer environment, hiding the problem. It will definitely cause PR-time (or, worse, later!) headaches when a developer writes code that is supported by std::optional but not by our Optional and (if we are lucky!) leads to a compile-time error in some CI environment.

Since we have to write Squid code with the "weakest point" (i.e. our custom/incomplete Optional implementation) in mind anyway, what do we have to _gain_ by doing this substitution? A tiny bit of performance improvement in some std::optional implementations? Even that is not guaranteed and has not been demonstrated. I cannot think of enough reasons to justify the associated risks...

[^1]: I am not implying that the current or [proposed](https://github.com/squid-cache/squid/pull/1106) Optional replacement is buggy, but I know first-hand that it is easy to make difficult-to-discover bugs in that low-level code. We are operating at the edge of our abilities here and, hence, should be extra careful. The cost/benefit tradeoff here is different than in most other/legitimate compat/ cases!

Reply:
I have updated PR title/description to emphasize the important part of this PR changes. The move to libcompat is a boring technicality compared to the ramifications of using std::optional (instead of Optional) where the former is available. Please adjust further as needed.

Reply:
What we gain with this model of portability is that bugs in our custom code is restricted to affecting only outdated OS.

While this is not great for users of those OS, outdated OS only decrease in usage over time. Reducing the impact of any bugs present in the custom implementation(s). Also counteracting technical debt incurred by delays bug fixing.